src/transformers/generation_tf_utils.py

```diff
@@ -2260,6 +2305,511 @@ def sample(
+            # to existing finished scores and select the best from the new set of beams
+            sequences_seq_last = tf.transpose(sequences.stack(), perm=[1, 2, 0])
+            merged_sequences = tf.concat([sequences_seq_last, topk_sequences_seq_last], axis=1)
+            merged_scores = tf.concat([scores, topk_log_probs], axis=1)
+            merged_is_sent_finished = tf.concat([is_sent_finished, did_topk_just_finished], axis=1)
+            topk_merged_indices = tf.math.top_k(merged_scores, k=num_beams)[1]
+            next_sequences_seq_last, next_scores, next_is_sent_finished = gather_beams(
+                [merged_sequences, merged_scores, merged_is_sent_finished], topk_merged_indices
+            )
+
+            # 8. Update model kwargs.
+            # Determine the top k beam indices from the original set of all beams. With these, gather the top k
+            # beam-associated caches.
+            next_running_indices = gather_beams(topk_beam_indices, next_topk_indices)
+            next_cache = gather_beams(cache, next_running_indices, batch_axis=cache_batch_axis)
+            model_outputs["past_key_values"] = tf.nest.map_structure(
+                lambda tensor: flatten_beam_dim(tensor, batch_axis=cache_batch_axis), next_cache
+            )
+            if use_xla:
+                next_model_kwargs = self._update_model_kwargs_for_xla_generation(
+                    model_outputs, model_kwargs, cur_len, max_length
+                )
+            else:
+                next_model_kwargs = self._update_model_kwargs_for_generation(
+                    model_outputs, model_kwargs, is_encoder_decoder=self.config.is_encoder_decoder
+                )
+
+            # 9. Prepare the `tf.TensorArray` for the next iteration
+            next_sequences = sequences.unstack(tf.transpose(next_sequences_seq_last, perm=[2, 0, 1]))
+            next_running_sequences = running_sequences.unstack(
+                tf.transpose(next_running_sequences_seq_last, perm=[2, 0, 1])
+            )
+
+            return (
+                cur_len + 1,
+                next_running_sequences,
+                next_running_scores,
+                next_sequences,
+                next_scores,
+                next_is_sent_finished,
+                next_model_kwargs,
+            )
+
+        # 5. run generation
+        # Adds the `intermediary_running_sequences` TensorArray into the body, needed as a scratchpad
+        beam_search_body_fn = partial(
+            beam_search_body_fn, intermediary_running_sequences=intermediary_running_sequences
+        )
+
+        # 1st generation step has to be run before to initialize `past`
+        cur_len, running_sequences, running_scores, sequences, scores, is_sent_finished, model_kwargs = partial(
+            beam_search_body_fn, input_ids_length=input_ids.shape[-1]
+        )(cur_len, running_sequences, running_scores, sequences, scores, is_sent_finished, model_kwargs)
```

Comment:
I'm confused about why this is a `partial` created and called once instead of just a straight call to the function!

Reply:
No hard need, actually :) I did it mostly so the FLAX/TF body calls would be nearly identical, but can remove the partial

Reply:
Can we do the partial on a separate line? This is quite unreadable.